app/src/main/java/pl/futurecollars/invoicing/db/file/FileBasedInvoiceRepository.java

```diff
@@ -0,0 +1,140 @@
+package pl.futurecollars.invoicing.db.file;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import java.io.IOException;
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import pl.futurecollars.invoicing.db.InvoiceRepository;
+import pl.futurecollars.invoicing.exceptions.FileSystemException;
+import pl.futurecollars.invoicing.model.Invoice;
+import pl.futurecollars.invoicing.services.FileService;
+import pl.futurecollars.invoicing.services.JsonService;
+
+public class FileBasedInvoiceRepository implements InvoiceRepository {
+
+    private final FileService fileService;
+    private final JsonService<Invoice> jsonService;
+
+    public FileBasedInvoiceRepository(FileService fileService, JsonService<Invoice> jsonService) {
+        this.fileService = fileService;
+        this.jsonService = jsonService;
+    }
+
+    @Override
+    public Invoice save(Invoice invoice) {
+        UUID id = UUID.randomUUID();
+        try {
+            if (!this.getById(id).isPresent()) {
+                invoice.setId(id);
+                String json = jsonService.toJson(invoice);
+                fileService.writeToFile(json);
+                return invoice;
+            } else {
+                return save(invoice);
```

Comment:
czy tutaj czasem jakaś rekurencja nie powstaje? :D Jeżeli istnieje już Invoice z takim ID to wywołasz tą samą metodę (save) ponownie. Ona znowu sprawdzi "a, istnieje invoice z takim id no to wywołuję save(..)" i tak w nieskończoność.

O wiele lepiej byłoby  rzucić exception (ID zajęty)